sui_core/src/authority_active/gossip/mod.rs

```diff
@@ -47,40 +47,72 @@ where
         degree,
     );
 
+    // If we do not expect to connect to anyone
+    if target_num_tasks == 0 {
+        info!("Turn off gossip mechanism");
+        return;
+    }
+    info!("Turn on gossip mechanism");
+
     // Keep track of names of active peers
     let mut peer_names = HashSet::new();
     let mut gossip_tasks = FuturesUnordered::new();
 
     // TODO: provide a clean way to get out of the loop.
     loop {
+        debug!("Seek new peers");
+
+        // Find out what is the earliest time that we are allowed to reconnect
+        // to at least 2f+1 nodes.
+        let next_connect = active_authority
+            .minimum_wait_for_majority_honest_available()
+            .await;
+        debug!(
+            "Waiting for {:?}",
+            next_connect - tokio::time::Instant::now()
+        );
+        tokio::time::sleep_until(next_connect).await;
+
         let mut k = 0;
         while gossip_tasks.len() < target_num_tasks {
             let name = active_authority.state.committee.sample();
-            if peer_names.contains(name) || *name == active_authority.state.name {
+            if peer_names.contains(name)
+                || *name == active_authority.state.name
+                || !active_authority.can_contact(*name).await
+            {
                 continue;
             }
             peer_names.insert(*name);
             gossip_tasks.push(async move {
                 let peer_gossip = PeerGossip::new(*name, active_authority);
                 // Add more duration if we make more than 1 to ensure overlap
-                info!("Gossip: Start gossip from peer {:?}", *name);
+                debug!("Start gossip from peer {:?}", *name);
                 peer_gossip
                     .spawn(Duration::from_secs(REFRESH_FOLLOWER_PERIOD_SECS + k * 15))
                     .await
             });
             k += 1;
+
+            // Only try to connect 20 times, after that re-assess whether
+            // there are enough people to connect to.
+            if k > 20 {
```

Comment:
Make 20 a constant?

Reply:
I actually added a proper check here instead of a hack: now we check if we already have used 2/3 stake in terms of who we follow, and exit the loop if we do. And wait for someone to be unfollowed to look again.